genie-agent/src/main/java/com/netflix/genie/agent/execution/services/impl/grpc/GRpcJobTerminationListenerServiceImpl.java

```diff
@@ -0,0 +1,118 @@
+/*
+ *
+ *  Copyright 2018 Netflix, Inc.
+ *
+ *     Licensed under the Apache License, Version 2.0 (the "License");
+ *     you may not use this file except in compliance with the License.
+ *     You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *     Unless required by applicable law or agreed to in writing, software
+ *     distributed under the License is distributed on an "AS IS" BASIS,
+ *     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *     See the License for the specific language governing permissions and
+ *     limitations under the License.
+ *
+ */
+package com.netflix.genie.agent.execution.services.impl.grpc;
+
+import com.google.common.util.concurrent.FutureCallback;
+import com.google.common.util.concurrent.Futures;
+import com.netflix.genie.agent.execution.ExecutionContext;
+import com.netflix.genie.agent.execution.services.JobTerminationListenerService;
+import com.netflix.genie.proto.JobKillListenerRequest;
+import com.netflix.genie.proto.JobKillListenerResponse;
+import com.netflix.genie.proto.JobTerminationServiceGrpc;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.context.annotation.Lazy;
+import org.springframework.stereotype.Service;
+
+import javax.annotation.Nullable;
+import javax.validation.constraints.NotNull;
+
+/**
+ * Implementation of the {@link JobTerminationListenerService}.
+ *
+ * @author standon
+ * @since 4.0.0
+ */
+@Lazy
+@Service
+@Slf4j
+public class GRpcJobTerminationListenerServiceImpl implements JobTerminationListenerService {
+
+    private final JobTerminationServiceGrpc.JobTerminationServiceFutureStub client;
+    private final ExecutionContext executionContext;
+
+    /**
+     * Constructor.
+     *
+     * @param client           The gRPC client to use to call the server.
+     * @param executionContext The execution context for the agent
+     */
+    public GRpcJobTerminationListenerServiceImpl(
+        final JobTerminationServiceGrpc.JobTerminationServiceFutureStub client,
+        final ExecutionContext executionContext
+    ) {
+        this.client = client;
+        this.executionContext = executionContext;
+    }
+
+
+    @Override
+    public void registerForRemoteKillNotification(@NotNull final Runnable onKillCallback) {
+        final String jobId = executionContext.getClaimedJobId();
+
+        Futures.addCallback(
+            this.client
+                .registerForKillNotification(
+                    JobKillListenerRequest.newBuilder()
+                        .setJobId(jobId)
+                        .build()
+                ),
+            new JobKillFutureCallback(onKillCallback)
+        );
+    }
+
+    private void handleErrorListeningForKillNotification(final Runnable onSuccessCallback) {
+        registerForRemoteKillNotification(onSuccessCallback);
```

Comment:
This seems like a possible infinite loop

Reply:
Yup. Had discussed this with @mprimi. Instead of a retry policy we thought to have this as the initial implementation